proposals/0000-provenance-qualified-imports.rst

```diff
@@ -0,0 +1,107 @@
+
+    source-origin-aliases:
+       hackage: https://…
+       foobar: http://…
+
+``build-depends:`` fields now can take package names in the form of ``[REPO]+somepackage`` rather than just ``somepackage``.
+
+Further, there is an implicitly defined ``baserepo`` alias that always refers to the base repo set by the user in their ``~/.cabal/config`` (or in the command-line, if the config file is overridden).
+
+Effect and Interactions
+-----------------------
+I can think of no difficult interactions with existing features. There may need to be a fair amount of mechanical work to remove assumptions about uniqueness of package names throughout different portions of the codebase. I do not believe that there is any interaction with backpack, but it would be important to confirm this.
+
+Costs and Drawbacks
+-------------------
+Aliases may be considered confusing and/or unnecessary, although I think they will be a boon to usability.
+
+More importantly, there is a concern that widespread use of provenance information could lead to fragmentation of the ecosystem, brittle dependencies, etc. I think this is largely resolved by two choices that this proposal recommends making.
+
+1) ``cabal sdist`` should emit a warning on use of the source-origin field.
+2) the central hackage server should prevent upload of packages making use of ``source-origin`` in their ``.cabal`` file.
+
+Alternatives
+------------
+The only current choice when package names overlap is to fork and rename one of the overlapping packages, which is unfortunate.
+
+Possible areas of disagreement/future work
+--------------------
+1) The syntax chosen to separate provenance information from a package name is the ``+`` character. This was chosen because ``/`` and ``:`` are present in too many urls (and the latter may be used for distinguishing subpackages as well, in the future). There may be other views as to the best syntax here. One other suggestion would be a literal space (`` ``) as this is not legal in either urls or package names.
+
+2) The question arises: should there be some global location for alias maps? At this point, I think we should not have this. In particular, the aim of the proposal is to allow genuine federation, without the need of any central authority. By introducing some global location by which names are assigned (outside of the mechanisms for assigning names that are common to the internet as a whole), we find ourselves with another version of the same problem. In the future, if we find it is too awkward for packages to each assign aliases, this can be revisited. (Also note -- it is expected that the use of these aliases is not going to be common, to begin with, and will generally not be for _central_ packages, but rather, for the most part, for proprietary code, and individual applications).
```

Comment:
```suggestion
2) The question arises: should there be some global location for alias maps? At this point, I think we should not have this. In particular, the aim of the proposal is to allow genuine federation, without the need of any central authority. By introducing some global location by which names are assigned (outside of the mechanisms for assigning names that are common to the internet as a whole), we find ourselves with another version of the same problem. In the future, if we find it is too awkward for packages to each assign aliases, this can be revisited. (Also note -- it is expected that the use of these aliases is not going to be common, to begin with, and will generally not be for *central* packages, but rather, for the most part, for proprietary code, and individual applications).
```